### DISCUSSION TITLE
Attribute-driven min feature zoom

### DISCUSSION THREAD

**Discussion by msbarry** [maplibre/maplibre-gl-js#1867] [Ideas]
I have a use-case where I want to set a fractional "min zoom" attribute on individual features so as you zoom in and out, the feature is visible above that zoom and invisible below it. I don't think there's a way to do this in off-the-shelf maplibre so in the past I've done this in a fork by tweaking the shader but I'm wondering if it would make sense to add as a feature to maplibre.

Ideally I'd be able to do something like:

```javascript
'circle-minzoom': ['get', 'minzoom']
```

and that minzoom would get included in the feature's data passed to the gpu and the shader would handle conditionally showing/hiding the feature by comparing to zoom scalar.

A few things I tried using existing style spec:

```javascript
'circle-opacity': [
  'case',
  ['>=', ['zoom'], ['get', 'minzoom']], 1,
  0
]
```

but it fails with error:

> Error: layers.point.paint.circle-opacity: "zoom" expression may only be used as input to a top-level "step" or "interpolate" expression.

or:

```javascript
'circle-opacity': [
  'step', ["zoom"],
  0,
  ["get", "minzoom"], 1
]
```

but it fails with:

> Input/output pairs for "step" expressions must be defined using literal numeric values (not computed expressions) for the input values.

Does anyone have any other ideas to try for doing this with existing features? Or if not then what do people think of the proposed `*-minzoom` style spec addition?

**Comment by HarelM** [comment]
We are doing this in the "data level" i.e. when building the tiles we select which features to include in the tiles and which not to include, why not solve it this way?

**Reply by msbarry** [reply]
Yep, I'm doing that as well, but I need to be able to show/hide individual features at precise, fractional zoom levels.

**Reply by HarelM** [reply]
Can you share a short movie to better explain what you mean?

**Reply by pnorman** [reply]
If you want a feature to appear at zoom 14.5, it must be in the zoom 14 tiles. So you have to filter it out until you reach 14.5. As discussed, there's no obvious good way to do it right now.

If you're at higher zooms where you're using overzoom, you also need to rely on style-side filtering, but that mostly works if you only need integer zoom level filtering.

**Reply by HarelM** [reply]
I might be able to imagine it, but if this was implemented in a fork it should be fairly easy to add a video here to align expectation :-)

**Reply by msbarry** [reply]
Here's an example with text showing up at zoom=1.5 and zoom=1.75.  The console is logging current zoom level on each "zoom" event.

https://user-images.githubusercontent.com/1480504/203171278-bcaa5561-855d-4291-9211-e063dd1c7952.mov

**Reply by msbarry** [reply]
Does that make sense @HarelM ?  I've only implemented it for text but I suppose it could make sense for any feature.

**Reply by wipfli** [reply]
I read somewhere that also other properties are only evaluated at integer zoom levels. Do you know what features are affected @msbarry?

**Reply by msbarry** [reply]
Any "layout property" from https://maplibre.org/maplibre-gl-js-docs/style-spec/layers/ would only be evaluated at integer zoom levels. Paint properties vary within zoom levels, but the ways in which they can vary by both data and zoom are limited and not quite flexible enough to support this use-case (from what I can see at least).

**Reply by HarelM** [reply]
Thanks for the video! It helps a lot.
I would expect this to work somehow, but it might be that some things are not properly supported.
Instead of defining new spec property nanes, I think the right direction would be to enhance the support of the existing style language to the relevant use case you have, much like some of the things that were suggested in the slack channel.
What do you think?

**Reply by msbarry** [reply]
That sounds good, do you have a sense which of the ones I tried (or a different one) would be ideal to try to get working? My only concern would be if there's some reason deep down that makes one of those options not viable...

**Reply by HarelM** [reply]
I honestly don't know. I would guess that these were not implemented only due to relevant use case not tested, not anything deeper.
I would suggest to post here a few possibilities and see which one the people here think is the best and try and implement it. If there's a problem with implementation, I would then go and try implementing the second favorite option. But up to you how to proceed with this, this is just my personal suggestion on how to move forward.
Thanks for all the time and effort you put into this! Truly appreciated.

**Reply by pnorman** [reply]
I suspect layout properties are only evaluated at integer zooms for performance reasons and for stability of features as you zoom in and out.

**Reply by msbarry** [reply]
From what I can see to get step/interpolate working with data-driven zoom stops you would need to compute the stops then pass them down in a uniform and compare them in the shader, and each shader would need support for each kind of attribute you can control with this mechanism.

Another option would be to just implement enough support for very basic minzoom/maxzoom cases using the existing step syntax. So the cases of `['step', ['zoom'], 0, ['get', 'minzoom'], 1]` or `['step', ['zoom'], 1, ['get', 'minzoom'], 0]` would go from being invalid to supported.

**Reply by HarelM** [reply]
I would consider adding support for the use case you need and not "for everything", unless it's almost the same code-wise which doesn't sound like it from what you described.
Having said that, I would try and avoid defining new syntax or keywords and stick with what currently exists, hopefully.
I hope that makes sense...

**Comment by wipfli** [comment]
@maxammann and I were discussing the other day that expressions are evaluated on the CPU and only at integer zoom levels. The interpolation at the fractional zoom levels is then done on the GPU. I guess this is for performance optimization.

@maxammann you said that you can confirm that the shaders only do the interpolation?

**Reply by IvanSanchez** [reply]
> expressions are evaluated on the CPU and only at integer zoom levels. The interpolation at the fractional zoom levels is then done on the GPU. I guess this is for performance optimization.

AFAIK *layout* expressions are evaluated in CPU, whereas *paint* expressions are evaluated in the GPU (in the shaders).

Layout stuff defines what data goes into vertex attributes into the GPU, then paint stuff defines how those attributes are evaluated into a colour. It's not really a matter of performance, but rather of *what can be done*.

**Reply by wipfli** [reply]
Ah interesting, so the full expressions for paint properties are evaluated on the GPU?

**Reply by IvanSanchez** [reply]
As far as I'm aware, yes.

Keeping in mind that my knowledge of maplibre's internals is lacking, I'm guessing that the original problem is that maplibre-gl-js has incomplete support for translating paint expressions into shader code.

**Reply by maxammann** [reply]
Expressions are always evaluated on the CPU. With evaluation I mean a function of `fn evaluate(expression: JSON) -> number | fn interpolate(min: number, max: number: zoom: number) -> number`.

The trick maplibre uses to make expression dependent on the zoom is that they evaluate the returned function on the GPU. That function is only interpolating linearly between two values though.

maplibre does not translate expressions to shader code in any way.

---
Bottom line: It's hard to say that expressions are only evaluated on the GPU or CPU. Render tricks are all over the place and make sure that expressions do not impact the render performance.

**Reply by maxammann** [reply]
Here is some more background information: https://maxammann.github.io/maplibre-internals/architecture/Expression-Architecture.html

**Reply by wipfli** [reply]
Cool

**Comment by msbarry** [comment]
I ended up going with the fork approach for speed of implementation, here's a demo of what the result looks like comparing each feature against it's minzoom to decide whether to show it in the shader: https://msbarry.github.io/maplibre-minzoom-demo

**Reply by wipfli** [reply]
Feels really fast!

**Reply by wipfli** [reply]
Do you have the code somewhere on a branch for MapLibre GL JS?

**Reply by wipfli** [reply]
And is there a comparison when `symbol-sort-key` or so would be use and collision detection would be done with normal MapLibre GL JS? Is it much slower?

**Reply by maxammann** [reply]
Very interesting!

**Reply by msbarry** [reply]
I made https://msbarry.github.io/maplibre-minzoom-demo/client.html for comparison for a client-side approach.  It's also a different feel since maplibre's symbol placement  tries to fill empty space more aggressively with lower-priority labels, which disappear when you zoom in and higher-priority ones take over. Maplibre's collision avoidance probably makes an individual view look better, but as a user I often find labels disappearing when you zoom into them kind of annoying.

**Reply by wipfli** [reply]
Oh wau, my phone struggles with rendering the normal version. Feels like less than 1 frame per second

**Reply by msbarry** [reply]
and just for fun, here's the same approach applied to 486k placenames from the whole planet with labels bucketed into zoom levels and served using pmtiles: https://msbarry.github.io/maplibre-minzoom-demo/world.html

**Reply by wipfli** [reply]
Such a cool map

**Reply by wipfli** [reply]
@msbarry can you maybe share the source code of your attribute-driven minzoom implementation? I would like to test it with my map style...

**Reply by msbarry** [reply]
Sure thing - here's the diff: https://github.com/msbarry/maplibre-gl-js/commit/bf3b8de80957b934c4ab415174db9a702ee37a97

**Reply by wipfli** [reply]
Thanks!

**Comment by adjourn** [comment]
I ran into similar situation.

Since the MVP (_most valuable player_, not _minimum viable product_) of data generation `planetiler` has max zoom 15, there's a lot of stuff there and a big portion of it should be shown at zoom 16 (or even 17). Obvious solution with that limitation is to include `minzoom` attribute to features on zoom level 15 which I'm definitely going to do.

While I could benefit from other features as well, most obvious one would be `text-minzoom` for me instead of messing around with `visibility` or `text-opacity`.

The unfortunate outcome is that all zoom levels pay the price, although the `minzoom` variable exists only on zoom level 15.

**Reply by adjourn** [reply]
Okay, time to laugh at myself, took me few hours to experience what was already discussed here.. You CAN'T use zoom to match feature `minzoom` property in order to set `text-opacity/visibility`. And it seems you can't set layer `minzoom` dynamically.

Or has there been any advances in style spec to somehow achieve it?
